node-cpp.md

```diff
@@ -1,66 +1,115 @@
+1. To port a C++ project to Node to expose a new interface for the tool (like Mapnik & Node Mapnik)
+1. Improve performance at scale where Node becomes the bottleneck (i.e. concurrency)
 
-More examples of how to port C++ libraries to node can be found at [nodejs.org/api/addons](http://nodejs.org/api/addons.html). See https://nodesource.com/blog/c-add-ons-for-nodejs-v4/ for a detailed summary of the origins of Nan. And see http://blog.reverberate.org/2016/10/17/native-extensions-memory-management-part2-javascript-v8.html for a good introduction to v8 that is not specific to node.js.
+**Concurrency**
+
+Concurrency is the process of executing different pieces of the same process to allow for parallel execution of these pieces [[wikipedia](https://en.wikipedia.org/wiki/Concurrency_(computer_science))]. Node.js addons allow us to take advantage of concurrent operations within our node applications by reaching into more than just the v8 thread, but this can result in a few surprises. In the example below, we’ve passed data from our node application, into the v8 thread, and subsequently into our worker threadpool. Here there are multiple workers executing code at the same time. In our worker, we have the following line:
+
+```
+std::cout << "royal with cheese" << std::endl;
+```
 
-### Why create an addon?
+This will print `royal with cheese` to the terminal. In an application running this multiple times we’d expect the output to look like this:
 
-An addon is a viable solution for the following reasons:
+```
+royal with cheese
+royal with cheese
+royal with cheese
+royal with cheese
+// ... and so on
+```
 
-1. To port a C++ project to Node to expose a new interface for the tool (like Mapnik & Node Mapnik)
-1. Improve performance at scale where Node becomes the bottleneck.
+![not concurrent](https://mapbox.s3.amazonaws.com/cpp-assets/addon-hey-nonconcurrent.gif)
+
+But when we start printing from within the threadpool all of these can start executing simultaneously.
+
+![concurrent](https://mapbox.s3.amazonaws.com/cpp-assets/addon-hey-concurrent.gif)
+
+We run the following script node index.js. Here’s the output:
+
+```
+royal with cheese
+royraolry roawoyliya talwhl i  wtcwihhit ethceh hs ceeche
+heseeeres
+oseye
+r
+aolry oawrylioa tylwha i lwtc ihhwt eihcet hshcee he
+cesheeres
+oeeys
+raeol
+ry oawrylioa tylwha i lwtc ihhwt eihcet hshcee he
+cesheeres
+oeeys
+raeol
+ry oawrylioa tylwha i lwtc ihhwt eihcet hshcee he
+cesheeres
+oeeys
+raeol
+```
+
+That’s a messy burger! 🍔 std::cout is logging at the same time and space is being filled concurrently. We’re literally seeing concurrency happen here–awesome! This makes testing pretty hard though. There are a few options to get your logs in order, such as adding a mutex to keep things straight. Since we’re just testing, we can tell our computer to only run this script with a single thread - forcing our application to run non-concurrently. We pass the following before running our script:
```

Comment:
Link to [mutex](http://en.cppreference.com/w/cpp/thread/mutex) (not in our glossary yet)